### ISSUE TITLE
HardSigmoid on VPU

### ISSUE THREAD

**Issue by lino-alves** [openvinotoolkit/openvino#13014]
Hi,

I have a NN model in ONNX (converted from PyTorch), which uses HardSigmoid, which apparently is not supported by the VPU plugin.
I read a bit about the extensibility mechanism and it appears it can solve my problem if I replace the HardSigmoid by its basic operations (add, clamp and divide).
Is it possible to create a shared library using the ConversionExtension, like you do with the OpExtension? Are there any C++ examples using the ConversionExtension like you have for the OpExtension, which is quite complete?

Also, is there any example for Python using the ConversionExtension to add the extension before reading the model?

Thanks for the support.

**Comment by ilya-lavrenov** [comment]
Hi @lino-alves 
Since HardSigmoid is already part of OpenVINO opset, you can just use usual transformations https://docs.openvino.ai/latest/openvino_docs_transformations.html to convert hardsigmoid to a set of operations.
ConversionExtension are useful when an operation is not a part of OpenVINO opset and we need to convert it during model reading from unknown operation to a set of OV supported operations.

If you managed to implement such transformation, it would be nice to see it as PR to MYRIAD plugin

**Comment by slyalin** [comment]
Hi @lino-alves,

As a short-term solution you can try to use `ConversionExtension` for this purpose. But more correct way is to contribute transformation to the transformation library or to the the plugin itself and call it inside the plugin transformation pipeline as @ilya-lavrenov suggested. The later requires more knowledge from your side though.

Yes, you can put `ConversionExtension` to a shared library, I've just checked this with https://github.com/openvinotoolkit/openvino/blob/master/docs/template_extension/new/ov_extension.cpp by adding a new entry of type `ov::frontend::ConversionExtensionton` to the vector of extensions like this:

```
OPENVINO_CREATE_EXTENSIONS(
    std::vector<ov::Extension::Ptr>({

        // Register operation itself, required to be read from IR
        std::make_shared<ov::OpExtension<TemplateExtension::Identity>>(),

        // Register operaton mapping, required when converted from framework model format
        std::make_shared<ov::frontend::OpExtension<TemplateExtension::Identity>>(),

        // Dummy example of replacing Relu by Identity custom op
        std::make_shared<ov::frontend::ConversionExtension>("Relu", [](const ov::frontend::NodeContext& node) {
            return std::make_shared<TemplateExtension::Identity>(node.get_input(0))->outputs();
        })
    }));
```
It is loaded then and applied during ONNX model conversion when you call add_extension before ONNX model read according to the documentation. Registering of ConversionExtension is the same as for OpExtension. Documentation gives examples of ConversionExtension here: https://docs.openvino.ai/latest/openvino_docs_Extensibility_UG_Frontend_Extensions.html#mapping-to-multiple-operations-with-conversionextension. But we still need more example here as you mentioned.

As for Python side of the documentation, there is a clear gap here. Usage is like this:

```
from openvino.runtime import Core
from openvino.frontend import ConversionExtension
import openvino.runtime.opset8 as ops

ov = Core()

# Dummy example of Relu(x) replaced by Add(x, x)
def conversion (node):
    return ops.add(node.get_input(0), node.get_input(0)).outputs()

ov.add_extension(ConversionExtension('Relu', conversion))

converted = ov.read_model('your_model.onnx')
```

**Comment by lino-alves** [comment]
Hi @slyalin,

Thanks a lot for the comprehensive examples, they helped me a lot.
I agree that updating the transformation suggested by @ilya-lavrenov would be the better way, but that means I would need to go deeper into the source code and I fear it could take me a while, so for now I am taking the fast way with ConversionExtension to have a first build of my code running as soon as possible and then later I can take a look at the transformation API when I am not under time pressure.
Having said that, I spent some time today working on this and apparently I managed to do the following:

1. In C++ I created a shared library with the extension for the HardSigmoid (HardSigmoid operation replacement implementation is just the Identity operation for now, I will finish this later);
2. Run model optimizer with the extension and verified HardSigmoid was replaced with my operation in Netron
3. Tried to run the inference in Python by loading the shared library before reading the OpenVINO IR (.xml). At this point I get an error: `RuntimeError: Cannot create HardSigmoidRep layer 644 id:119 from unsupported opset: extension`

In my extension file I have only:
`OPENVINO_CREATE_EXTENSIONS(

    std::vector<ov::Extension::Ptr>({
        // std::make_shared<ov::OpExtension<TemplateExtension::HardSigmoid>>(),
        std::make_shared<ov::frontend::ConversionExtension>("HardSigmoid", [](const ov::frontend::NodeContext& node) {
            return std::make_shared<TemplateExtension::HardSigmoid>(node.get_input(0))->outputs();
        })
    }));`
If I understand correctly this only gives the conversion for the frontend, so ONNX. For OpenVINO IR, I also need to register the operation in a similar way as in the example: `std::make_shared<ov::OpExtension<TemplateExtension::HardSigmoid>>()`. I have this line commented in my code above.
Is this flow correct and possible? Using a shared library with model optimizer and load the OpenVINO IR with the same library with ConversionExtension?

**Comment by slyalin** [comment]
@lino-alves, by uncommenting that line with OpExtension you can you the same library for both Model Optimizer and the run-time for your workflow. So Model Optimizer will use ConversionExtension while conversion to get know how to convert HardSigmoid your way and how to store it to IR. And then the runtime will use OpExtension to understand how to load it from IR and which real operation to use for in-memory representation and inference.

By the way you don't need really an extra pass of Model Optimizer because the runtime can load ONNX file directly for you. In this case you don't need to have separate OpExtension, ConversionExtension will do the job for you. If you still want to convert model first and then load it separately, both extensions are required if you are using a custom op class like TemplateExtension::HardSigmoid, which is not a big deal I think.

TemplateExtension::HardSigmoid is for your experiments with the extensibility, correct? I believe that what you needed initially was just ConversionExtension because all operations in your decomposition were standard one. So for your original request you don't need to touch OpExtension at all.

**Comment by lino-alves** [comment]
@slyalin, I understand that I could load the ONNX file directly, but I would like to avoid that, because I am using the Model Optimizer to compress the model to float16 (and a few more operations). If I load the IR model then I have everything in place with the Model Optimizer parameters.
Sorry that I didn't make this clear on my original post, where I only mention the ONNX file and never talked about the IR model file.

As you know I started working on the shared library for the extension and currently I have couple of points to ask you about.
First regarding the implementation in the evaluate function. Here's my current code:
```
bool HardSigmoid::evaluate(ov::TensorVector& outputs, const ov::TensorVector& inputs) const {
    auto in = inputs[0];
    auto out = outputs[0];

    auto calc = std::make_shared<ov::opset8::Maximum>(ov::opset8::Constant::create(ov::element::f16, ov::Shape{1}, {0}),
                std::make_shared<ov::opset8::Minimum>(ov::opset8::Constant::create(ov::element::f16, ov::Shape{1}, {1}),
                std::make_shared<ov::opset8::Add>(std::make_shared<ov::opset8::Multiply>(
                    ov::opset8::Constant::create(ov::element::f16, ov::Shape{1}, {alpha}), in), ov::opset8::Constant::create(ov::element::f16, ov::Shape{1}, {beta}))
                ));

    out.set_shape(in.get_shape());
    memcpy(out.data(), in.data(), in.get_byte_size());
    return true;
}
```
This is essentially the code from the Template and I added the operations to replace the HardSigmoid, however the function is not complete yet, because I failed to find how to pass the input tensor to the operation (`Multiply `in this case) and also how to pass the results of the operations on the variable `calc` to the output tensor. Currently I left the Template code in the function and the `in` variable in the `calc` operations is just a placeholder.
Unfortunately I am not as experienced in C++ as I am in Python or C, so any hints you can provide here are highly appreciated.

The second point is about the attributes. I added `alpha` and `beta` to the class definition:
```
class HardSigmoid : public ov::op::Op {
    float alpha = 0.2;
    float beta = 0.5;
```
I also added these to the `visit_attributes` function. I am however unsure if this is correct because when I inspect the IR model with Netron I see the new replacement operation with the attributes with the default values, instead of taking the values in the ONNX model. Is this correct?

**Comment by lino-alves** [comment]
I managed to solve this problem, but with a different solution than I initially had in mind.

Coming back to the extension, I spent some time working with the ConversionExtension and OpExtension to create an extension that would allow me to load the OpenVINO IR module to the VPU.
I managed to fix the problem previously described regarding the attributes, by passing them, if available, to the ConversionExtension, if not available the default value was passed. I verified with Netron that the attributes were correct.
I was however unable to code the class, since I was not able to glue the arguments from the evaluate function with the OpenVINO operations. I went through the objects from both the evaluate functions arguments and the OpenVINO operations and didn't find any way to pass the tensors from one to the other.
The ConversionExtension works perfectly if used standalone, i.e. without going through the evaluate function, this I tested both in C++ and Python. Unfortunately to use the OpenVINO IR models I need more than this.

Next I looked at the model optimizer. I thought about creating an extension to the model optimizer code to decompose the HardSigmoid. It would solve my problem in that I would have an OpenVINO IR model without the HardSigmoid that I could load to the VPU. However before doing that I tried something even easier. Since my original model is from PyTorch, I tweaked the ONNX exporter to export the HardSigmoid decomposed. This new ONNX model allowed me to run it through the model optimizer without any extensions and load the resulting OpenVINO IR to the VPU.
Surprisingly the model optimizer fuses the operations of the decomposed HardSigmoid into a HSigmoid, which the VPU supports. So the optimal solution for this issue would be to create an extension to the model optimizer that detects the HardSigmoid and replaces it by HSigmoid if the attribute values match (the alpha and beta from HardSigmoid match the fixed values in HSigmoid), however I am not sure about the feasibility of this because a float comparison would be needed, since in this case alpha is 1/6.

In summary, the easy solution is to output the ONNX model decomposed and then model optimizer will fuse the operations back together, the optimal solution would be to extend the model optimizer if feasible.

**Comment by zrnupping** [comment]
@lino-alves  Hello, I also meet this problem. If possible, can you share your code about how to "output the ONNX model decomposed and then model optimizer will fuse the operations back together"? Thanks.

**Comment by lino-alves** [comment]
@zrnupping You need to decompose the HardSigmoid operation in PyTorch's ONNX operations.
In file `torch.onnx.symbolic_opset9.py` change the `hardsigmoid` method from:
```
def hardsigmoid(g, self):
    return g.op('HardSigmoid', self, alpha_f=1 / 6)
```
to (for example)
```
def hardsigmoid(g, self):
    add = g.op("Add", self, g.op('Constant', value_t=torch.tensor(3, dtype=torch.float)))
    clip = g.op("Clip", add, g.op('Constant', value_t=torch.tensor(0, dtype=torch.float)), g.op('Constant', value_t=torch.tensor(6, dtype=torch.float)))
    div = g.op("Div", clip, g.op('Constant', value_t=torch.tensor(6, dtype=torch.float)))
    return div
```

**Comment by slyalin** [comment]
@lino-alves, just to wrap-up conversion story, here is the summary:

1. The correct way to fix it is to change VPU plugin to use existing transformations. Everything else is a kind of workaround.
2. Your attempts to impalement evaluate for `HardSigmoid` doesn't contribute to the story of supporting it on VPU. Evaluate function is a reference code for operation that would run on CPU only, VPU is not involved. Here is only a conversion extension should be involved, this is a type of extension created with `ov::frontend::ConversionExtension` or with `ov::frontend::OpExtension`. The purpose of these classes is to convert operation from Framework (from ONNX in your case) to OpenVINO operation. In our case we already have HardSigmoid as OpenVINO operation (https://docs.openvino.ai/latest/openvino_docs_ops_activation_HardSigmoid_1.html), so creating a new operation is not needed (so you don't need to use core `ov::OpExtension`).
3. If you can confirm that decomposition of HardSigmoid works at ONNX level, so the same code should work at the level of `ConversinoExtension`, you just need to put equivalent decomposition using OV ops. It can be done purely in Python as my example above illustrates. But if it works at Pytorch onnx export level, I would suggest to stay with this solution and wait until VPU plugin fix the issue.

**Comment by lino-alves** [comment]
I agree with the summary. I believe this issue can be closed.

**Comment by avitial** [comment]
Closing this. Feel free to reopen to ask any questions related to this topic.
